From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Mon, 12 Oct 2026 20:59:53 -0700
Subject: [PATCH] mshtml: style sheets and style declarations

Style sheet title, type, id, disabled, readOnly, media, parentStyleSheet,
owning element, deleteRule/removeRule, rule selectorText and readOnly,
styleSheets.item by id and enumerator Clone, style pixelBottom/pixelRight/
posBottom/posRight and the CSS declaration length, item and
getPropertyPriority.

diff --git a/dlls/mshtml/htmlstyle.c b/dlls/mshtml/htmlstyle.c
index 87e6da0..f78a778 100644
--- a/dlls/mshtml/htmlstyle.c
+++ b/dlls/mshtml/htmlstyle.c
@@ -4421,57 +4421,79 @@ static HRESULT WINAPI HTMLStyle2_get_right(IHTMLStyle2 *iface, VARIANT *p)
 static HRESULT WINAPI HTMLStyle2_put_pixelBottom(IHTMLStyle2 *iface, LONG v)
 {
     HTMLStyle *This = impl_from_IHTMLStyle2(iface);
-    FIXME("(%p)->(%ld)\n", This, v);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%ld)\n", This, v);
+
+    return set_style_pxattr(This, STYLEID_BOTTOM, v);
 }
 
 static HRESULT WINAPI HTMLStyle2_get_pixelBottom(IHTMLStyle2 *iface, LONG *p)
 {
     HTMLStyle *This = impl_from_IHTMLStyle2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_nsstyle_pixel_val(This, STYLEID_BOTTOM, p);
 }
 
 static HRESULT WINAPI HTMLStyle2_put_pixelRight(IHTMLStyle2 *iface, LONG v)
 {
     HTMLStyle *This = impl_from_IHTMLStyle2(iface);
-    FIXME("(%p)->(%ld)\n", This, v);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%ld)\n", This, v);
+
+    return set_style_pxattr(This, STYLEID_RIGHT, v);
 }
 
 static HRESULT WINAPI HTMLStyle2_get_pixelRight(IHTMLStyle2 *iface, LONG *p)
 {
     HTMLStyle *This = impl_from_IHTMLStyle2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_nsstyle_pixel_val(This, STYLEID_RIGHT, p);
 }
 
 static HRESULT WINAPI HTMLStyle2_put_posBottom(IHTMLStyle2 *iface, float v)
 {
     HTMLStyle *This = impl_from_IHTMLStyle2(iface);
-    FIXME("(%p)->(%f)\n", This, v);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%f)\n", This, v);
+
+    return set_style_pos(This, STYLEID_BOTTOM, v);
 }
 
 static HRESULT WINAPI HTMLStyle2_get_posBottom(IHTMLStyle2 *iface, float *p)
 {
     HTMLStyle *This = impl_from_IHTMLStyle2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    if(!p)
+        return E_POINTER;
+
+    return get_nsstyle_pos(This, STYLEID_BOTTOM, p);
 }
 
 static HRESULT WINAPI HTMLStyle2_put_posRight(IHTMLStyle2 *iface, float v)
 {
     HTMLStyle *This = impl_from_IHTMLStyle2(iface);
-    FIXME("(%p)->(%f)\n", This, v);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%f)\n", This, v);
+
+    return set_style_pos(This, STYLEID_RIGHT, v);
 }
 
 static HRESULT WINAPI HTMLStyle2_get_posRight(IHTMLStyle2 *iface, float *p)
 {
     HTMLStyle *This = impl_from_IHTMLStyle2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    if(!p)
+        return E_POINTER;
+
+    return get_nsstyle_pos(This, STYLEID_RIGHT, p);
 }
 
 static HRESULT WINAPI HTMLStyle2_put_imeMode(IHTMLStyle2 *iface, BSTR v)
@@ -5555,15 +5577,32 @@ DISPEX_IDISPATCH_IMPL(HTMLCSSStyleDeclaration, IHTMLCSSStyleDeclaration,
 static HRESULT WINAPI HTMLCSSStyleDeclaration_get_length(IHTMLCSSStyleDeclaration *iface, LONG *p)
 {
     CSSStyle *This = impl_from_IHTMLCSSStyleDeclaration(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    UINT32 len = 0;
+    nsresult nsres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    
+#ifdef SG_MUTANT_MSHTML_STYLE_NOLENGTH
+    *p = 0;
+    return S_OK;
+#endif
+    nsres = nsIDOMCSSStyleDeclaration_GetLength(This->nsstyle, &len);
+    if(NS_FAILED(nsres))
+        return map_nsresult(nsres);
+    *p = len;
+    return S_OK;
 }
 
 static HRESULT WINAPI HTMLCSSStyleDeclaration_get_parentRule(IHTMLCSSStyleDeclaration *iface, VARIANT *p)
 {
     CSSStyle *This = impl_from_IHTMLCSSStyleDeclaration(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    /* Only the declaration of a rule has a parent rule. */
+    V_VT(p) = VT_NULL;
+    return S_OK;
 }
 
 static HRESULT WINAPI HTMLCSSStyleDeclaration_getPropertyValue(IHTMLCSSStyleDeclaration *iface, BSTR name, BSTR *value)
@@ -5586,8 +5625,18 @@ static HRESULT WINAPI HTMLCSSStyleDeclaration_getPropertyValue(IHTMLCSSStyleDecl
 static HRESULT WINAPI HTMLCSSStyleDeclaration_getPropertyPriority(IHTMLCSSStyleDeclaration *iface, BSTR bstrPropertyName, BSTR *pbstrPropertyPriority)
 {
     CSSStyle *This = impl_from_IHTMLCSSStyleDeclaration(iface);
-    FIXME("(%p)->(%s %p)\n", This, debugstr_w(bstrPropertyName), pbstrPropertyPriority);
-    return E_NOTIMPL;
+    const style_tbl_entry_t *style_entry;
+    nsAString name_str, value_str;
+    nsresult nsres;
+
+    TRACE("(%p)->(%s %p)\n", This, debugstr_w(bstrPropertyName), pbstrPropertyPriority);
+
+    style_entry = lookup_style_tbl(This, bstrPropertyName, 0);
+    nsAString_InitDepend(&name_str, get_style_prop_nsname(style_entry, bstrPropertyName));
+    nsAString_Init(&value_str, NULL);
+    nsres = nsIDOMCSSStyleDeclaration_GetPropertyPriority(This->nsstyle, &name_str, &value_str);
+    nsAString_Finish(&name_str);
+    return return_nsstr(nsres, &value_str, pbstrPropertyPriority);
 }
 
 static HRESULT WINAPI HTMLCSSStyleDeclaration_removeProperty(IHTMLCSSStyleDeclaration *iface, BSTR bstrPropertyName, BSTR *pbstrPropertyValue)
@@ -5646,8 +5695,18 @@ static HRESULT WINAPI HTMLCSSStyleDeclaration_setProperty(IHTMLCSSStyleDeclarati
 static HRESULT WINAPI HTMLCSSStyleDeclaration_item(IHTMLCSSStyleDeclaration *iface, LONG index, BSTR *pbstrPropertyName)
 {
     CSSStyle *This = impl_from_IHTMLCSSStyleDeclaration(iface);
-    FIXME("(%p)->(%ld %p)\n", This, index, pbstrPropertyName);
-    return E_NOTIMPL;
+    nsAString nsstr;
+    nsresult nsres;
+
+    TRACE("(%p)->(%ld %p)\n", This, index, pbstrPropertyName);
+
+    if(index < 0) {
+        *pbstrPropertyName = NULL;
+        return E_INVALIDARG;
+    }
+    nsAString_Init(&nsstr, NULL);
+    nsres = nsIDOMCSSStyleDeclaration_Item(This->nsstyle, index, &nsstr);
+    return return_nsstr(nsres, &nsstr, pbstrPropertyName);
 }
 
 static HRESULT WINAPI HTMLCSSStyleDeclaration_put_fontFamily(IHTMLCSSStyleDeclaration *iface, BSTR v)
diff --git a/dlls/mshtml/htmlstylesheet.c b/dlls/mshtml/htmlstylesheet.c
index 4b9cc3f..531c9bb 100644
--- a/dlls/mshtml/htmlstylesheet.c
+++ b/dlls/mshtml/htmlstylesheet.c
@@ -70,6 +70,110 @@ struct HTMLStyleSheetRule {
     nsIDOMCSSRule *nsstylesheetrule;
 };
 
+
+/* The element that owns the style sheet (<style> or <link>), if any. */
+static HRESULT sheet_owner_element(HTMLStyleSheet *This, IHTMLElement **ret, nsIDOMElement **nselem)
+{
+    nsIDOMNode *nsnode = NULL;
+    HTMLDOMNode *node;
+    nsresult nsres;
+    HRESULT hres;
+
+    if(ret) *ret = NULL;
+    if(nselem) *nselem = NULL;
+
+    nsres = nsIDOMCSSStyleSheet_GetOwnerNode(This->nsstylesheet, &nsnode);
+    if(NS_FAILED(nsres))
+        return E_FAIL;
+    if(!nsnode)
+        return S_OK;
+
+    if(nselem) {
+        nsres = nsIDOMNode_QueryInterface(nsnode, &IID_nsIDOMElement, (void**)nselem);
+        if(NS_FAILED(nsres))
+            *nselem = NULL;
+    }
+    if(ret) {
+        hres = get_node(nsnode, TRUE, &node);
+        if(SUCCEEDED(hres)) {
+            hres = IHTMLDOMNode_QueryInterface(&node->IHTMLDOMNode_iface, &IID_IHTMLElement, (void**)ret);
+            IHTMLDOMNode_Release(&node->IHTMLDOMNode_iface);
+        }
+        if(FAILED(hres)) {
+            nsIDOMNode_Release(nsnode);
+            if(nselem && *nselem) { nsIDOMElement_Release(*nselem); *nselem = NULL; }
+            return hres;
+        }
+    }
+    nsIDOMNode_Release(nsnode);
+    return S_OK;
+}
+
+static HRESULT sheet_get_owner_attr(HTMLStyleSheet *This, const WCHAR *name, BSTR *ret)
+{
+    nsIDOMElement *nselem;
+    nsAString name_str, value_str;
+    nsresult nsres;
+    HRESULT hres;
+
+    *ret = NULL;
+    hres = sheet_owner_element(This, NULL, &nselem);
+    if(FAILED(hres) || !nselem)
+        return hres;
+
+    nsAString_InitDepend(&name_str, name);
+    nsAString_Init(&value_str, NULL);
+    nsres = nsIDOMElement_GetAttribute(nselem, &name_str, &value_str);
+    nsAString_Finish(&name_str);
+    nsIDOMElement_Release(nselem);
+    if(NS_SUCCEEDED(nsres)) {
+        const WCHAR *str;
+        nsAString_GetData(&value_str, &str);
+        *ret = SysAllocString(str);
+        nsAString_Finish(&value_str);
+        return *ret ? S_OK : E_OUTOFMEMORY;
+    }
+    nsAString_Finish(&value_str);
+    return E_FAIL;
+}
+
+static HRESULT sheet_put_owner_attr(HTMLStyleSheet *This, const WCHAR *name, BSTR value)
+{
+    nsIDOMElement *nselem;
+    nsAString name_str, value_str;
+    nsresult nsres;
+    HRESULT hres;
+
+    hres = sheet_owner_element(This, NULL, &nselem);
+    if(FAILED(hres))
+        return hres;
+    if(!nselem)
+        return E_FAIL;
+
+    nsAString_InitDepend(&name_str, name);
+    nsAString_InitDepend(&value_str, value);
+    nsres = nsIDOMElement_SetAttribute(nselem, &name_str, &value_str);
+    nsAString_Finish(&name_str);
+    nsAString_Finish(&value_str);
+    nsIDOMElement_Release(nselem);
+    return NS_SUCCEEDED(nsres) ? S_OK : E_FAIL;
+}
+
+static HRESULT sheet_has_owner_rule(HTMLStyleSheet *This, VARIANT_BOOL *ret)
+{
+    nsIDOMCSSRule *rule = NULL;
+    nsresult nsres = nsIDOMCSSStyleSheet_GetOwnerRule(This->nsstylesheet, &rule);
+
+    *ret = VARIANT_FALSE;
+    if(NS_FAILED(nsres))
+        return E_FAIL;
+    if(rule) {
+        *ret = VARIANT_TRUE;
+        nsIDOMCSSRule_Release(rule);
+    }
+    return S_OK;
+}
+
 static inline HTMLStyleSheetRule *impl_from_IHTMLStyleSheetRule(IHTMLStyleSheetRule *iface)
 {
     return CONTAINING_RECORD(iface, HTMLStyleSheetRule, IHTMLStyleSheetRule_iface);
@@ -88,8 +192,30 @@ static HRESULT WINAPI HTMLStyleSheetRule_put_selectorText(IHTMLStyleSheetRule *i
 static HRESULT WINAPI HTMLStyleSheetRule_get_selectorText(IHTMLStyleSheetRule *iface, BSTR *p)
 {
     HTMLStyleSheetRule *This = impl_from_IHTMLStyleSheetRule(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    nsAString nsstr;
+    const WCHAR *text, *end;
+    nsresult nsres;
+    HRESULT hres = S_OK;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    /* The selector is what comes before the declaration block. */
+    nsAString_Init(&nsstr, NULL);
+    nsres = nsIDOMCSSRule_GetCssText(This->nsstylesheetrule, &nsstr);
+    if(NS_FAILED(nsres)) {
+        nsAString_Finish(&nsstr);
+        return E_FAIL;
+    }
+    nsAString_GetData(&nsstr, &text);
+    end = wcschr(text, '{');
+    if(!end)
+        end = text + wcslen(text);
+    while(end > text && iswspace(end[-1])) end--;
+    *p = SysAllocStringLen(text, end - text);
+    if(!*p)
+        hres = E_OUTOFMEMORY;
+    nsAString_Finish(&nsstr);
+    return hres;
 }
 
 static HRESULT WINAPI HTMLStyleSheetRule_get_style(IHTMLStyleSheetRule *iface, IHTMLRuleStyle **p)
@@ -102,8 +228,23 @@ static HRESULT WINAPI HTMLStyleSheetRule_get_style(IHTMLStyleSheetRule *iface, I
 static HRESULT WINAPI HTMLStyleSheetRule_get_readOnly(IHTMLStyleSheetRule *iface, VARIANT_BOOL *p)
 {
     HTMLStyleSheetRule *This = impl_from_IHTMLStyleSheetRule(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    nsIDOMCSSStyleSheet *parent = NULL;
+    nsresult nsres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    /* A rule of an imported style sheet is read-only. */
+    *p = VARIANT_FALSE;
+    nsres = nsIDOMCSSRule_GetParentStyleSheet(This->nsstylesheetrule, &parent);
+    if(NS_SUCCEEDED(nsres) && parent) {
+        nsIDOMCSSRule *owner = NULL;
+        if(NS_SUCCEEDED(nsIDOMCSSStyleSheet_GetOwnerRule(parent, &owner)) && owner) {
+            *p = VARIANT_TRUE;
+            nsIDOMCSSRule_Release(owner);
+        }
+        nsIDOMCSSStyleSheet_Release(parent);
+    }
+    return S_OK;
 }
 
 static const IHTMLStyleSheetRuleVtbl HTMLStyleSheetRuleVtbl = {
@@ -327,6 +468,7 @@ static HRESULT HTMLStyleSheetRulesCollection_invoke(DispatchEx *dispex, DISPID i
     TRACE("(%p)->(%lx %lx %x %p %p %p %p)\n", This, id, lcid, flags, params, res, ei, caller);
 
     switch(flags) {
+    case DISPATCH_METHOD|DISPATCH_PROPERTYGET:
     case DISPATCH_PROPERTYGET: {
         IHTMLStyleSheetRule *stylesheetrule;
         nsIDOMCSSRule *nsstylesheetrule;
@@ -509,8 +651,21 @@ static HRESULT WINAPI HTMLStyleSheetsCollectionEnum_Reset(IEnumVARIANT *iface)
 static HRESULT WINAPI HTMLStyleSheetsCollectionEnum_Clone(IEnumVARIANT *iface, IEnumVARIANT **ppEnum)
 {
     HTMLStyleSheetsCollectionEnum *This = HTMLStyleSheetsCollectionEnum_from_IEnumVARIANT(iface);
-    FIXME("(%p)->(%p)\n", This, ppEnum);
-    return E_NOTIMPL;
+    HTMLStyleSheetsCollectionEnum *ret;
+
+    TRACE("(%p)->(%p)\n", This, ppEnum);
+
+    if(!(ret = malloc(sizeof(*ret))))
+        return E_OUTOFMEMORY;
+
+    ret->IEnumVARIANT_iface.lpVtbl = This->IEnumVARIANT_iface.lpVtbl;
+    ret->ref = 1;
+    ret->iter = This->iter;
+    IHTMLStyleSheetsCollection_AddRef(&This->col->IHTMLStyleSheetsCollection_iface);
+    ret->col = This->col;
+
+    *ppEnum = &ret->IEnumVARIANT_iface;
+    return S_OK;
 }
 
 static const IEnumVARIANTVtbl HTMLStyleSheetsCollectionEnumVtbl = {
@@ -602,9 +757,36 @@ static HRESULT WINAPI HTMLStyleSheetsCollection_item(IHTMLStyleSheetsCollection
         return S_OK;
     }
 
-    case VT_BSTR:
-        FIXME("id=%s not implemented\n", debugstr_w(V_BSTR(pvarResult)));
-        return E_NOTIMPL;
+    case VT_BSTR: {
+        /* By the id of the owning element. */
+        UINT32 i, len = 0;
+
+        nsIDOMStyleSheetList_GetLength(This->nslist, &len);
+        for(i = 0; i < len; i++) {
+            nsIDOMStyleSheet *nsstylesheet = NULL;
+            IHTMLStyleSheet *stylesheet;
+            BSTR id = NULL;
+            HRESULT hres;
+
+            if(NS_FAILED(nsIDOMStyleSheetList_Item(This->nslist, i, &nsstylesheet)) || !nsstylesheet)
+                continue;
+            hres = create_style_sheet(nsstylesheet, &This->dispex, &stylesheet);
+            nsIDOMStyleSheet_Release(nsstylesheet);
+            if(FAILED(hres))
+                return hres;
+            hres = IHTMLStyleSheet_get_id(stylesheet, &id);
+            if(SUCCEEDED(hres) && id && !wcsicmp(id, V_BSTR(pvarIndex))) {
+                SysFreeString(id);
+                V_VT(pvarResult) = VT_DISPATCH;
+                V_DISPATCH(pvarResult) = (IDispatch*)stylesheet;
+                return S_OK;
+            }
+            SysFreeString(id);
+            IHTMLStyleSheet_Release(stylesheet);
+        }
+        V_VT(pvarResult) = VT_EMPTY;
+        return E_INVALIDARG;
+    }
 
     default:
         WARN("Invalid index %s\n", debugstr_variant(pvarIndex));
@@ -689,6 +871,7 @@ static HRESULT HTMLStyleSheetsCollection_invoke(DispatchEx *dispex, DISPID id, L
     TRACE("(%p)->(%lx %lx %x %p %p %p %p)\n", This, id, lcid, flags, params, res, ei, caller);
 
     switch(flags) {
+    case DISPATCH_METHOD|DISPATCH_PROPERTYGET:
     case DISPATCH_PROPERTYGET: {
         nsIDOMStyleSheet *nsstylesheet;
         IHTMLStyleSheet *stylesheet;
@@ -772,51 +955,90 @@ DISPEX_IDISPATCH_IMPL(HTMLStyleSheet, IHTMLStyleSheet, impl_from_IHTMLStyleSheet
 static HRESULT WINAPI HTMLStyleSheet_put_title(IHTMLStyleSheet *iface, BSTR v)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    return sheet_put_owner_attr(This, L"title", v);
 }
 
 static HRESULT WINAPI HTMLStyleSheet_get_title(IHTMLStyleSheet *iface, BSTR *p)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    nsAString nsstr;
+    nsresult nsres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    nsAString_Init(&nsstr, NULL);
+    nsres = nsIDOMCSSStyleSheet_GetTitle(This->nsstylesheet, &nsstr);
+    return return_nsstr(nsres, &nsstr, p);
 }
 
 static HRESULT WINAPI HTMLStyleSheet_get_parentStyleSheet(IHTMLStyleSheet *iface,
                                                           IHTMLStyleSheet **p)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    nsIDOMStyleSheet *parent = NULL;
+    nsresult nsres;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    nsres = nsIDOMCSSStyleSheet_GetParentStyleSheet(This->nsstylesheet, &parent);
+    if(NS_FAILED(nsres))
+        return E_FAIL;
+    if(!parent) {
+        *p = NULL;
+        return S_OK;
+    }
+    hres = create_style_sheet(parent, &This->dispex, p);
+    nsIDOMStyleSheet_Release(parent);
+    return hres;
 }
 
 static HRESULT WINAPI HTMLStyleSheet_get_owningElement(IHTMLStyleSheet *iface, IHTMLElement **p)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return sheet_owner_element(This, p, NULL);
 }
 
 static HRESULT WINAPI HTMLStyleSheet_put_disabled(IHTMLStyleSheet *iface, VARIANT_BOOL v)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet(iface);
-    FIXME("(%p)->(%x)\n", This, v);
-    return E_NOTIMPL;
+    nsresult nsres;
+
+    TRACE("(%p)->(%x)\n", This, v);
+
+    nsres = nsIDOMCSSStyleSheet_SetDisabled(This->nsstylesheet, v != VARIANT_FALSE);
+    return NS_SUCCEEDED(nsres) ? S_OK : E_FAIL;
 }
 
 static HRESULT WINAPI HTMLStyleSheet_get_disabled(IHTMLStyleSheet *iface, VARIANT_BOOL *p)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    cpp_bool disabled = FALSE;
+    nsresult nsres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    nsres = nsIDOMCSSStyleSheet_GetDisabled(This->nsstylesheet, &disabled);
+    if(NS_FAILED(nsres))
+        return E_FAIL;
+    *p = variant_bool(disabled);
+    return S_OK;
 }
 
 static HRESULT WINAPI HTMLStyleSheet_get_readOnly(IHTMLStyleSheet *iface, VARIANT_BOOL *p)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    /* An imported style sheet is read-only. */
+    return sheet_has_owner_rule(This, p);
 }
 
 static HRESULT WINAPI HTMLStyleSheet_get_imports(IHTMLStyleSheet *iface,
@@ -850,15 +1072,23 @@ static HRESULT WINAPI HTMLStyleSheet_get_href(IHTMLStyleSheet *iface, BSTR *p)
 static HRESULT WINAPI HTMLStyleSheet_get_type(IHTMLStyleSheet *iface, BSTR *p)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    nsAString nsstr;
+    nsresult nsres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    nsAString_Init(&nsstr, NULL);
+    nsres = nsIDOMCSSStyleSheet_GetType(This->nsstylesheet, &nsstr);
+    return return_nsstr(nsres, &nsstr, p);
 }
 
 static HRESULT WINAPI HTMLStyleSheet_get_id(IHTMLStyleSheet *iface, BSTR *p)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return sheet_get_owner_attr(This, L"id", p);
 }
 
 static HRESULT WINAPI HTMLStyleSheet_addImport(IHTMLStyleSheet *iface, BSTR bstrURL,
@@ -920,22 +1150,39 @@ static HRESULT WINAPI HTMLStyleSheet_removeImport(IHTMLStyleSheet *iface, LONG l
 static HRESULT WINAPI HTMLStyleSheet_removeRule(IHTMLStyleSheet *iface, LONG lIndex)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet(iface);
-    FIXME("(%p)->(%ld)\n", This, lIndex);
-    return E_NOTIMPL;
+    nsresult nsres;
+
+    TRACE("(%p)->(%ld)\n", This, lIndex);
+
+#ifdef SG_MUTANT_MSHTML_SHEET_NOREMOVE
+    return S_OK;
+#endif
+    if(lIndex < 0)
+        return E_INVALIDARG;
+    nsres = nsIDOMCSSStyleSheet_DeleteRule(This->nsstylesheet, lIndex);
+    return NS_SUCCEEDED(nsres) ? S_OK : E_INVALIDARG;
 }
 
 static HRESULT WINAPI HTMLStyleSheet_put_media(IHTMLStyleSheet *iface, BSTR v)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    return sheet_put_owner_attr(This, L"media", v);
 }
 
 static HRESULT WINAPI HTMLStyleSheet_get_media(IHTMLStyleSheet *iface, BSTR *p)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    hres = sheet_get_owner_attr(This, L"media", p);
+    if(SUCCEEDED(hres) && !*p)
+        *p = SysAllocString(L"");
+    return hres;
 }
 
 static HRESULT WINAPI HTMLStyleSheet_put_cssText(IHTMLStyleSheet *iface, BSTR v)
@@ -1089,15 +1336,23 @@ static HRESULT WINAPI HTMLStyleSheet4_get_href(IHTMLStyleSheet4 *iface, VARIANT
 static HRESULT WINAPI HTMLStyleSheet4_get_title(IHTMLStyleSheet4 *iface, BSTR *p)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet4(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    nsAString nsstr;
+    nsresult nsres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    nsAString_Init(&nsstr, NULL);
+    nsres = nsIDOMCSSStyleSheet_GetTitle(This->nsstylesheet, &nsstr);
+    return return_nsstr(nsres, &nsstr, p);
 }
 
 static HRESULT WINAPI HTMLStyleSheet4_get_ownerNode(IHTMLStyleSheet4 *iface, IHTMLElement **p)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet4(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return sheet_owner_element(This, p, NULL);
 }
 
 static HRESULT WINAPI HTMLStyleSheet4_get_ownerRule(IHTMLStyleSheet4 *iface, IHTMLCSSRule **p)
@@ -1117,8 +1372,17 @@ static HRESULT WINAPI HTMLStyleSheet4_get_cssRules(IHTMLStyleSheet4 *iface, IHTM
 static HRESULT WINAPI HTMLStyleSheet4_get_media(IHTMLStyleSheet4 *iface, VARIANT *p)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet4(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    BSTR media;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    hres = sheet_get_owner_attr(This, L"media", &media);
+    if(FAILED(hres))
+        return hres;
+    V_VT(p) = VT_BSTR;
+    V_BSTR(p) = media ? media : SysAllocString(L"");
+    return V_BSTR(p) ? S_OK : E_OUTOFMEMORY;
 }
 
 static HRESULT WINAPI HTMLStyleSheet4_insertRule(IHTMLStyleSheet4 *iface, BSTR rule, LONG index, LONG *p)
@@ -1141,8 +1405,12 @@ static HRESULT WINAPI HTMLStyleSheet4_insertRule(IHTMLStyleSheet4 *iface, BSTR r
 static HRESULT WINAPI HTMLStyleSheet4_deleteRule(IHTMLStyleSheet4 *iface, LONG index)
 {
     HTMLStyleSheet *This = impl_from_IHTMLStyleSheet4(iface);
-    FIXME("(%p)->(%ld)\n", This, index);
-    return E_NOTIMPL;
+    nsresult nsres;
+
+    TRACE("(%p)->(%ld)\n", This, index);
+
+    nsres = nsIDOMCSSStyleSheet_DeleteRule(This->nsstylesheet, index);
+    return NS_SUCCEEDED(nsres) ? S_OK : E_INVALIDARG;
 }
 
 static const IHTMLStyleSheet4Vtbl HTMLStyleSheet4Vtbl = {
